### ISSUE TITLE
Object method named 'dict' conflicts with class attribute of type dict

### ISSUE THREAD

**Issue by declaresub** [python/cpython#103121]
Here is what appears to be a bug:

>>> class M:
...     def dict(self):
...         pass
...     foo: dict[str, str] = {}
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 4, in M
TypeError: 'function' object is not subscriptable

The class M is created, but without a dict method. One can replace 'dict' by 'list' or 'set' and generate the same exception.  

If the type hint for foo is omitted, then all is well.

>>> class M:
...     def dict(self):
...         pass
...     foo = {}
... 

Python version: 3.11.2, Mac OS 13.3,  Apple M1 Max.

**Comment by gaogaotiantian** [comment]
Are you sure class `M` is created? I belive since you are using the interactive shell, the class M you defined previously is still left in the local namespace and that's what you saw.

I tried with a clean shell on 3.11.2:
```
Python 3.11.1 (main, Dec  7 2022, 01:11:44) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class M:
...     def dict(self):
...             pass
...     foo: dict[str, str] = {}
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 4, in M
TypeError: 'function' object is not subscriptable
>>> M
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'M' is not defined
```

This is expected behavior as you redefined variable `dict` in your class definition.

**Comment by gresm** [comment]
Let's simplify (it doesn't work exactly like this, but close enough):

Syntax for creating the class is as follows:
```py
class [ClassName]:
    [code_block]
    ...
```

What really happens is that python:
1) creates an execution scope (for local variables and such)
2) executes the code in the ``[code_block]``
3) checks what local variables were changed and using them, it builds the class

So, what really happens is that you truly replace the built-in ``dict`` object with a function named ``dict``. And while the built-in class ``dict`` defines a magic method ``__class_getitem__``, the **function** ``dict`` doesn't. What is ``__class_getitem__``? It is a class method that is called when using ``ClassName[some_value]``.  Other thing you might not know, is that annotations don't really have a special syntax:
```py
[var_name]: [code_expression] = [code_expression]
```
The code block that makes a type annotation is actually executed and the result saved in ``[class].__annotations__`` (if it is an annotation for that class member or in ``__annotations__`` (if it is an annotation in a global scope).

So, your code could be interpreted like this:
```py
def some_random_func():
    pass

dict = some_random_func

dict[some_value]  # raises a TypeError, because functions don't support such an operation
```

The latter error ``name 'M' is not defined`` is caused by the fact that when the error was raised in the previous part, the class creation was aborted, so no variable named "M" existed when trying to access it.

In summary: this isn't a python bug, but really a mix of intentional features, that mixed together in this confusing way, seemed like a bug.

**Comment by gresm** [comment]
Adding to the above comment:
Changing the line order should work.
```py
class M:
    foo: dict[str, str] = {}
    def dict(self):
        pass
```

As the given code can be interpreted like this:

```py
dict[some_value]  # works now , because this is an original built-in ``dict`` class

def some_random_func():
    pass

dict = some_random_func
```

And as @gaogaotiantian said, please reconsider your function name, as it overrides the built-in ``dict`` that you or other people could think that is left intact.

**Comment by TeamSpen210** [comment]
Another approach is to `import builtins`, then access `dict` from there. But indeed it's better to avoid using builtin names for things wherever possible, due to confusing results like this.

**Comment by sobolevn** [comment]
This is a completely expected behavior.
Scoping rules are applied to annotations as well.
In the class scope `M`, `dict` is redefined to be a method.
This is why it fails with `TypeError: 'function' object is not subscriptable`

If you want to fix this, please use type aliases.

**Comment by declaresub** [comment]
I think I now understand why this is not a bug, but a gotcha.  Thanks for the explanations.
